### PR TITLE
Update readme

### PR DESCRIPTION
**Pull request by mattrunyon**
This at least adds some info about development scripts/workflow for Python/Java. Also removes the bulk of the main readme which was a copy our docs quickstart guide

There could still be more improvements such as adding info on how to do testing

Added placeholders for things I know have to be manually generated if there are protobuf changes.

### CODE REVIEW

File: proto/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Deephaven Protobuf
+
+For Java, proto changes will be automatically refelected when building/running the Java server app. Sections below list files which must be manually generated and committed to source control when proto files change.
```
Comment (margaretkennedy) [inline proto/README.md:3]:
```suggestion
For Java, proto changes will be automatically reflected when building/running the Java server app. Sections below list files that must be manually generated and committed to source control when proto files change.
```

File: proto/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+## JS API
+
+Needs info on how to generate
+
+## Python API
+
+Needs info on how to generate
```
Comment (margaretkennedy) [inline proto/README.md:11]: @chipkent ?

File: py/README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+# Deephaven Python Development
+
+## Requirements
+
+- Python >=3.8
+
+## Setting up Virtual Environment
+
+For any of the options below, you should set up a virtual environment in Python first. Once created, you must activate the virtual environment in each shell you open each time.
```
Comment (margaretkennedy) [inline py/README.md:9]:
```suggestion
For any of the options below, you should first set up a virtual environment in Python. Once created, you must activate the virtual environment every time you open a shell. 
```

File: py/README.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+```sh
+python3 -m venv .venv # Create. Only needed once
+source .venv/bin/activate # Activate the virtual environment
+deactivate # Exit virtual environment when done
+```
+
+**Note:**
+
+- This is not an exhaustive guide to managing python environments
```
Comment (margaretkennedy) [inline py/README.md:19]:
```suggestion
- This is not an exhaustive guide to managing Python environments.
```

File: py/README.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+- Depending on your OS and how your PATH is setup, you may need to use `python3`, or a path to the explicit python version you want to use
```
Comment (margaretkennedy) [inline py/README.md:20]:
```suggestion
- Depending on your OS and how your PATH is set up, you may need to use `python3`, or a path to the explicit Python version you want to use.
```

File: py/README.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+- You may choose to setup a "permanent" virtual environment location
```
Comment (margaretkennedy) [inline py/README.md:21]:
```suggestion
- You may set up a "permanent" virtual environment location.
```

File: py/README.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+- You'll need to re-install the wheel anytime you are making python code changes that affect the wheel
```
Comment (margaretkennedy) [inline py/README.md:22]:
```suggestion
- You'll need to re-install the wheel whenever you make Python code changes that affect the wheel.
```

File: py/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+- `pip` can be a pain if you are trying to (re-)install a wheel with the same version number as before
```
Comment (margaretkennedy) [inline py/README.md:23]:
```suggestion
- `pip` can be a pain if you are trying to (re-)install a wheel with the same version number as before.
```

File: py/README.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  - A `pip install --force-reinstall --no-deps "py/server/build/wheel/deephaven_core-<version>-py3-none-any.whl[autocomplete]"` may do the trick
```
Comment (margaretkennedy) [inline py/README.md:24]:
```suggestion
  - A `pip install --force-reinstall --no-deps "py/server/build/wheel/deephaven_core-<version>-py3-none-any.whl[autocomplete]"` may do the trick.
```

File: py/README.md [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+- You can install other python packages in your venv using `pip install <some-other-package>`
```
Comment (margaretkennedy) [inline py/README.md:25]:
```suggestion
- You can install other Python packages in your venv using `pip install <some-other-package>`.
```

File: py/README.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+- You can setup multiple virtual environments, and switch between them as necessary using `source /path/to/other-venv/bin/activate`
```
Comment (margaretkennedy) [inline py/README.md:26]:
```suggestion
- You can set up multiple virtual environments and switch between them as necessary using `source /path/to/other-venv/bin/activate`.
```

File: py/README.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+- You can de-activate the virtual environment by running `deactivate`
```
Comment (margaretkennedy) [inline py/README.md:27]:
```suggestion
- You can de-activate the virtual environment by running `deactivate`.
```

File: py/README.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+- You can use the `VIRTUAL_ENV` environment variable instead of sourcing / activating virtual environments: `VIRTUAL_ENV=/my/venv ./gradlew server-jetty-app:run`
```
Comment (margaretkennedy) [inline py/README.md:28]:
```suggestion
- You can use the `VIRTUAL_ENV` environment variable instead of sourcing/activating virtual environments: `VIRTUAL_ENV=/my/venv ./gradlew server-jetty-app:run`.
```

File: py/README.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+## Local Development
+
+It is recommended to use a pip editable install so that Python changes are reflected when your server is restarted. These instructions are for the Python server package in `py/server`, but should apply to any other Python packages.
```
Comment (margaretkennedy) [inline py/README.md:32]:
```suggestion
Using a `pip` editable install is recommended so that Python changes are reflected when your server is restarted. These instructions are for the Python server package in `py/server` but should apply to any other Python packages.
```

File: py/README.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+Install the Python server package. This will install the dependencies as well as the autocomplete dependencies.
+
+```sh
+DEEPHAVEN_VERSION=0.23.0 pip install -e "py/server[autocomplete]"
+```
+
+**Note:** Pip may report an error that the version is not what is expected if you have installed the locally built version before. Adjust the version to match what pip expects.
```
Comment (margaretkennedy) [inline py/README.md:40]:
```suggestion
**Note:** Pip may report an error that the version is not what is expected if you have installed the locally built version before. Adjust the version to match what `pip` expects.
```

File: py/README.md [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+After installing, you can run `pip list` and you should see the package `deephaven-core` with an editable project location listed.
+
+Now you can start the Java server by following the instructions [here](../server/jetty-app/README.md). Any time you make Python changes, restart the server to apply them.
+
+## Local Build
+
+You can use these instructions if you do not plan on making frequent changes to your local Python environment or need to try things in a more production-like environment locally.
+
+First run the following command to build the server and embedded-server if needed. The embedded-server is used to start [Deephaven from Python](https://deephaven.io/core/docs/how-to-guides/configuration/native-application/#python-embedded-server)
```
Comment (margaretkennedy) [inline py/README.md:50]:
```suggestion
First, run the following command to build the server and embedded-server if needed. The embedded-server is used to start [Deephaven from Python](https://deephaven.io/core/docs/how-to-guides/configuration/native-application/#python-embedded-server).
```

File: py/README.md [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+```sh
+pip install --force py/server/build/wheel/deephaven_core-<version>-py3-none-any.whl
+
+# If using embedded-server
+# You can combine as pip install core.whl server.whl if you need both
+pip install --force py/embedded-server/build/wheel/deephaven_server-<version>-py3-none-any.whl
+```
+
+Now you can start the Java server by following the instructions [here](../server/jetty-app/README.md)
```
Comment (margaretkennedy) [inline py/README.md:72]:
```suggestion
Start the Java server by following the instructions [here](../server/jetty-app/README.md).
```

File: proto/README.md (hunk 2 above)
Comment (mattrunyon) [reply proto/README.md]: Added the info in a followup commit

**Review by jmao-denver — APPROVED** [review]
LGTM

File: proto/README.md [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+./gradlew :py-client:updateProtobuf :go:updateProtobuf :cpp-client:updateProtobuf
+```
+
+## Authorization
+
+See [authorization readme](../authorization/README.md)
+
+## JS API
+
+Needs info on how to generate
```
Comment (margaretkennedy) [inline proto/README.md:17]: Noting that this and the tests section in the py file still need info.

File: .gitignore [hunk 19]
```diff
@@ -47,6 +47,9 @@ devopsLocal/
 pyclient/venv
 pyclient/dist
 pyclient/*.egg-info
+venv
+.venv
+*.egg-info
```
Comment (chipkent) [inline .gitignore:52]: Why were these removed?  I think the `venv` is there to handle venvs that pycharm creates, and the `egg-info` is for python files that are not checked in.

File: README.md [hunk 20]
```diff
@@ -32,269 +34,51 @@ which functions as the data backbone for prominent hedge funds, banks, and finan
 
 ## Run Deephaven
 
-This section is a quick start guide for running Deephaven from pre-built images.  Almost all users will want to run Deephaven using pre-built images.  It is the easiest way to deploy.  For detailed instructions, see [Launch Deephaven from pre-built images](https://deephaven.io/core/docs/tutorials/quickstart).
+If you want to use Deephaven, see our [Quick start for Docker guide](https://deephaven.io/core/docs/tutorials/quickstart).
 
-Developers interested in tinkering with and modifying source code should build from the source code. For detailed instructions on how to do this, see [Build and launch Deephaven](https://deephaven.io/core/docs/how-to-guides/launch-build).
+Developers interested in modifying source code should read on for development instructions.
```
Comment (chipkent) [inline README.md:39]: The link to the build docs on the website should not have been lost.
https://deephaven.io/core/docs/how-to-guides/launch-build/

File: README.md [hunk 21]
```diff
@@ -32,269 +34,51 @@ which functions as the data backbone for prominent hedge funds, banks, and finan
...
-docker version
-docker-compose version
-docker run hello-world
-```
+| Package | Version                       | OS           |
+| ------- | ----------------------------- | ------------ |
+| Java    | >=11, <20                     | All          |
+| Python  | >= 3.8                        | All          |
+| Windows | 10 (OS build 20262 or higher) | Only Windows |
+| WSL     | 2                             | Only Windows |
```
Comment (chipkent) [inline README.md:52]: This dependency list does not match what is on the website.
https://deephaven.io/core/docs/how-to-guides/launch-build/
Is python a hard requirement?  If not, that is not represented in either place.

File: README.md [hunk 22]
```diff
@@ -2,6 +2,8 @@
 
 ![Deephaven Data Labs Logo](docs/images/Deephaven_GH_Logo.svg)
 
+[Quick start for Docker user guide](https://deephaven.io/core/docs/tutorials/quickstart)
```
Comment (chipkent) [inline README.md:5]: We should also reference the quick start for pip.
https://deephaven.io/core/docs/tutorials/quickstart-pip/

File: README.md [hunk 23]
```diff
@@ -32,269 +34,51 @@ which functions as the data backbone for prominent hedge funds, banks, and finan
 
 ## Run Deephaven
 
-This section is a quick start guide for running Deephaven from pre-built images.  Almost all users will want to run Deephaven using pre-built images.  It is the easiest way to deploy.  For detailed instructions, see [Launch Deephaven from pre-built images](https://deephaven.io/core/docs/tutorials/quickstart).
+If you want to use Deephaven, see our [Quick start for Docker guide](https://deephaven.io/core/docs/tutorials/quickstart).
```
Comment (chipkent) [inline README.md:37]: We should also be directing people to https://deephaven.io/core/docs/tutorials/quickstart-pip/.  We need to provide very obvious flight paths for docker vs pip users.  This document is so lean that there is one flight path for end users.

File: README.md
Comment (chipkent) [inline README.md:1]: I completely agree that the readmes are due to be reworked.  Looking at this has pointed out a number of problems.  Overall, I think there are a number of things that need to be addressed:
1) These documents are ultra stripped down.  They do not provide enough context or meat to be good tools for outside contributors.
2) The documents are very stripped down "core dumps", rather than high quality external-facing documentation for all of our user types.  Basically, if you are already familiar with everything and know where to look, you can probably find what you are looking for.  The documents need more high level storyline and context to be generally useful.  For example, I don't think someone from DHE could use these to be quickly productive.
3) In becoming super stripped down, these readmes have lost all material that would sell someone on trying the project.  We need something.
4) We are not effectively directing users to the website for tutorials and other key user-facing stuff.
5) We aren't providing clear storylines that new users can follow.  e.g. python pip is not present.  Everything is way too focused only on developers of the DHC codebase without enough focused on the users.
6) There are inconsistencies between this readme and the deephaven.io docs.  We need a consistent message.

I'm going to ask @margaretkennedy to also get involved since I think there is a legitimate amount of work between the website and the readme to end up with something everyone will be happy with.

File: README.md [hunk 24]
```diff
@@ -32,269 +34,51 @@ which functions as the data backbone for prominent hedge funds, banks, and finan
...
-right = newTable(
-        intCol("DeptID", 31, 33, 34, 35),
-        stringCol("DeptName", "Sales", "Engineering", "Clerical", "Marketing"),
-        stringCol("Telephone", "(646) 555-0134", "(646) 555-0178", "(646) 555-0159", "(212) 555-0111")
-    )
-
-t = left.join(right, "DeptID", "DeptName,DeptTelephone=Telephone")
-```
-
-![alt_text](docs/images/ide_first_query.png "Deephaven IDE First Query")
```
Comment (chipkent) [inline README.md:277]: This is an example of material to get new users to consider the product -- which should not have been deleted.

File: server/jetty-app/README.md [hunk 25]
```diff
@@ -1,70 +1,29 @@
...
-   ```shell
-   python -m venv /tmp/my-dh-venv
-   source /tmp/my-dh-venv/bin/activate
-   ```
-
-1. Build and install wheel
-
-   ```shell
-   ./gradlew py-server:assemble
+If using Python, see [Python development guide](../../py/README.md) for setting up the Python environment.
```
Comment (chipkent) [inline server/jetty-app/README.md:7]: Someone coming to this is going to be totally confused.  Here Python devs are directed away, but below, it shows how to run python.

File: server/jetty-app/README.md [hunk 26]
```diff
@@ -1,70 +1,29 @@
...
 
-   # replace with the appropriate <version>
-   pip install "py/server/build/wheel/deephaven_core-<version>-py3-none-any.whl[autocomplete]"
+## Local Development
 
-   # To install without the optional `autocomplete` feature, run:
-   # pip install "py/server/build/wheel/deephaven_core-<version>-py3-none-any.whl"
-   ```
+`server-jetty-app:run` will incorporate local Java changes on each start. If you are not frequently changing Java code, see the next section.
```
Comment (chipkent) [inline server/jetty-app/README.md:11]: Without reading below, it will not be obvious to a user that this is a gradle target.

File: server/jetty-app/README.md [hunk 27]
```diff
@@ -1,70 +1,29 @@
...
-* You'll need to re-install the wheel anytime you are making python code changes that affect the wheel
-* `pip` can be a pain if you are trying to (re-)install a wheel with the same version number as before
-  * A `pip install --force-reinstall --no-deps "py/server/build/wheel/deephaven_core-<version>-py3-none-any.whl[autocomplete]"` may do the trick
-* You can install other python packages in your venv using `pip install <some-other-package>`
-* You can setup multiple virtual environments, and switch between them as necessary using `source /path/to/other-venv/bin/activate`
-* You can de-activate the virtual environment by running `deactivate`
-* You can use the `VIRTUAL_ENV` environment variable instead of sourcing / activating virtual environments: `VIRTUAL_ENV=/my/venv ./gradlew server-jetty-app:run`
+```shell
+./gradlew server-jetty-app:run # Python enigne by default
+./gradlew server-jetty-app:run -Pgroovy # Groovy engine
```
Comment (chipkent) [inline server/jetty-app/README.md:15]: These are normally referred to as sessions and not engines in the docs.

File: server/jetty-app/README.md [hunk 28]
```diff
@@ -1,70 +1,29 @@
...
+```
 
-### Start script
+## Development with Infrequent Changes
 
-To create a more production-like environment, you can create and invoke the start script instead of running via gradle:
+To create a more production-like environment, you can create and invoke the start script instead of running via gradle. This is faster if you need to often restart the server without making any changes to Java code (such as Python server development).
```
Comment (chipkent) [inline server/jetty-app/README.md:20]: Again, we are discussing Python, when Python people were directed away in an early step.

File: server/jetty-app/README.md [hunk 29]
```diff
@@ -1,70 +1,29 @@
 # server-jetty-app
 
-## Native development
-
 This document is oriented towards getting a server up for local development.
-If you are interested in running a native production release, please see
+If you are interested in running a native production release or to see other options for starting the server, please see
```
Comment (chipkent) [inline server/jetty-app/README.md:4]: No context is provided about what the difference in a dev env and a prod env are and what that means.

File: README.md [hunk 30]
```diff
@@ -32,269 +34,51 @@ which functions as the data backbone for prominent hedge funds, banks, and finan
...
+See the following READMEs for info about developing in various environments within the Deephaven core repository.
 
+- [Java](./server/jetty-app/README.md)
+- [Python](./py/README.md)
+- [Web](./web/client-ui/README.md)
+- [Go](./go/README.md)
+- [C++](./cpp-client/README.md)
```
Comment (chipkent) [inline README.md:64]: Needs an R link.

**Review by chipkent — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 19 above)
Comment (mattrunyon) [reply .gitignore:52]: These were added. And anything in `.gitignore` without a leading or middle slash will match at any level. So it is effectively the same as `**/venv/**` or "ignore any folder at any depth called `venv`". I could actually remove the `pyclient/venv` since it will match under this new item

File: README.md (hunk 21 above)
Comment (mattrunyon) [reply README.md:52]: Python isn't a hard requirement. Docker isn't either technically I think.

I'm also not sure the Java version is entirely correct here or the website. I _think_ gradle requires a more specific version to run, but not 100% sure (I had some issues in intellij b/c I had my gradle version set too new)

File: py/README.md [hunk 31]
```diff
@@ -0,0 +1,76 @@
...
+```
+
+**Note:**
+
+- This is not an exhaustive guide to managing Python environments.
+- Depending on your OS and how your PATH is set up, you may need to use `python3`, or a path to the explicit Python version you want to use.
+- You may set up a "permanent" virtual environment location.
+- You'll need to re-install the wheel whenever you make Python code changes that affect the wheel.
+- `pip` can be a pain if you are trying to (re-)install a wheel with the same version number as before.
+  - A `pip install --force-reinstall --no-deps "py/server/build/wheel/deephaven_core-<version>-py3-none-any.whl[autocomplete]"` may do the trick.
```
Comment (mofojed) [inline py/README.md:24]: Don't we also need to install the `py/server-embedded/.../deephaven_server` wheel? I've been doing that... but maybe I don't need to. :shrug:

Comment (mattrunyon) [reply py/README.md:24]: I've found it unnecessary if you're not using the pip installed version of DH (DHaaL I think we called it?)

This section was just moved from the jetty-app readme b/c it didn't make sense there IMO, but it does look like it was just a brain dump to get the info on the page

**Comment by mofojed** [comment]
Superceded by #4209 
@mattrunyon will make sure the changes from @jjbrosnan incorporate the developer development instructions
